h2/src/main/org/h2/util/ValueHashMap.java

```diff
@@ -174,6 +191,39 @@ public V get(Value key) {
         }
         return list;
     }
+    
+    public Iterable<Map.Entry<Value, V>> entries() {
+        return new EntryIterable();
+    }
+    
+    private final class EntryIterable implements Iterable<Map.Entry<Value, V>> {
+        @Override
+        public Iterator<Map.Entry<Value, V>> iterator() {
+            return new EntryIterator();
+        }
+    }
+
+    private final class EntryIterator implements Iterator<Map.Entry<Value, V>> {
```

Comment:
Java 7 does not have default methods, so `remove()` should be implemented too.

Reply:
?? If we run it an older version of java, nobody will call remove() since it doesn't exist yet. If we run it on a newer version of Java, it will throw UnsupportedOperationException, which is fine (for our purposes)

Reply:
Also `hasNext()` and `next()` methods violate the `Iterator` contract. The `next()` method should advance it to the next record, `hasNext()` can only inform whether the next invocation of `next()` return a value or throw `NoSuchElementException`. It is legal to invoke `hasNext()` multiple times without invocation of `next()` and state of the iterator should not be changed.

But in this implementation `hasNext()` method used to traverse across map instead. That's not good.

Reply:
> ?? If we run it an older version of java, nobody will call remove() since it doesn't exist yet.
This method is from JDK 1.2. We can't compile non-abstract iterators without `remove()` before Java 8. In Java 8 this method got default implementation.